sys/net/gnrc/routing/ipv6_auto_subnets/gnrc_ipv6_auto_subnets.c

```diff
@@ -213,6 +339,218 @@ void gnrc_ipv6_nib_rtr_adv_pio_cb(gnrc_netif_t *upstream, const ndp_opt_pi_t *pi
+    uint8_t idx_start = 0;
+    uint8_t subnets = gnrc_netif_numof() - 1;
+    uint8_t resend_tx = CONFIG_GNRC_IPV6_AUTO_SUBNETS_TX_RETRIES;
+
+    DEBUG("auto_subnets: %u local subnets\n", subnets);
+
+    if (subnets == 0) {
+        return NULL;
+    }
+
+    /* setup the message queue */
+    msg_init_queue(server_queue, SERVER_MSG_QUEUE_SIZE);
+
+    /* register server to receive messages from given port */
+    gnrc_netreg_entry_init_pid(&server, CONFIG_GNRC_IPV6_AUTO_SUBNETS_PORT, thread_getpid());
+    gnrc_netreg_register(GNRC_NETTYPE_UDP, &server);
+
+    while (1) {
+        msg_receive(&msg);
+
+        switch (msg.type) {
+        case GNRC_NETAPI_MSG_TYPE_RCV:
+        {
+            gnrc_pktsnip_t *pkt = msg.content.ptr;
+            auto_subnets_request_t *request = pkt->data;
+
+            /* check if we already got an announcement from that host */
+            int res = _get_my_l2addr_rank(_upstream, pkt);
+            if (res) {
+                /* calculate total number of subnets */
+                subnets += request->num_subnets;
+
+                DEBUG("auto_subnets: %u new remote subnets, total %u\n",
+                      request->num_subnets, subnets);
+
+                /* If other host is before us in order of MAC address, add
+                 * their subnets to our offset */
+                if (res > 0) {
+                    idx_start += request->num_subnets;
+                }
+            }
+
+            gnrc_pktbuf_release(pkt);
+            break;
+        }
+        case SERVER_MSG_TYPE_TIMEOUT:
+            if (resend_tx--) {
+                uint32_t timeout_us;
+                auto_subnets_request_t request = {
+                    .num_subnets = gnrc_netif_numof() - 1,
+                };
+
+                /* broadcast the number of subnets we want to create */
+                _send_udp(_upstream, &ipv6_addr_all_routers_link_local,
+                          CONFIG_GNRC_IPV6_AUTO_SUBNETS_PORT,
+                          &request, sizeof(request));
+
+                /* configure timeout for resend */
+                timeout_us = random_uint32_range(
+                                CONFIG_GNRC_IPV6_AUTO_SUBNETS_TIMEOUT_MS * US_PER_MS / 2,
+                                CONFIG_GNRC_IPV6_AUTO_SUBNETS_TIMEOUT_MS * US_PER_MS);
+                xtimer_set_msg(&timeout_timer, timeout_us, &timeout_msg, _server_pid);
+                DEBUG("auto_subnets: announce sent, next timeout in %" PRIu32 " µs\n", timeout_us);
+            } else {
+                /* config round done, configure subnets */
+                _configure_subnets(subnets, idx_start, _upstream, &_pio_cache);
+
+                /* start a new round of counting */
+                resend_tx = CONFIG_GNRC_IPV6_AUTO_SUBNETS_TX_RETRIES;
+                memset(l2addrs, 0, sizeof(l2addrs));
+                idx_start = 0;
+                subnets = gnrc_netif_numof() - 1;
+            }
```

Comment:
Could be moved to its own function (when some variables are made global which, given that they top-initialized in a thread with an endless loop, could make sense... doesn't hide required memory in the stack).

Reply:
I mean the stack is allocated anyway, I probably won't tune it by a few bytes.
IMHO it's cleaner to *not* have global variables but keep them where they are used.

Reply:
IMHO that makes it *less* readable, but fine.